In [323]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
import numpy as np

In [324]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [325]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [326]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [327]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
columns_list = df_market_data.columns.tolist()

# Scale columns
market_data_scale = StandardScaler().fit_transform(df_market_data)
market_data_scale

array([[ 5.08529366e-01,  4.93193071e-01,  7.72200433e-01,
         2.35459633e-01, -6.74950963e-02, -3.55953481e-01,
        -2.51636882e-01],
       [ 1.85445894e-01,  9.34445040e-01,  5.58692121e-01,
        -5.43409317e-02, -2.73482725e-01, -1.15759474e-01,
        -1.99352110e-01],
       [ 2.17739616e-02, -7.06336853e-01, -2.16804207e-02,
        -6.10301536e-02,  8.00452481e-03, -5.50246924e-01,
        -2.82060506e-01],
       [-4.07643829e-02, -8.10928066e-01,  2.49457974e-01,
        -5.03879651e-02, -3.73164019e-01, -4.58258816e-01,
        -2.95546142e-01],
       [ 1.19303608e+00,  2.00095907e+00,  1.76061001e+00,
         5.45842065e-01, -2.91202870e-01, -4.99847761e-01,
        -2.70316950e-01],
       [ 8.91870708e-01,  1.32729453e+00,  8.00214184e-01,
        -5.71478992e-02,  7.78653106e-01, -1.88231917e-01,
        -2.25532605e-01],
       [ 1.13972400e-02,  2.57225091e+00,  1.10164693e+00,
        -4.90495415e-01, -9.31954023e-01,  3.87758986e-01,
        -1.8284399

In [328]:
# Create a DataFrame with the scaled data
df_scaled_data = pd.DataFrame(market_data_scale, columns=columns_list)
df_scaled_data.head(10)

# Copy the crypto names from the original DataFrame
df_scaled_data["coin_id"] = df_market_data.index

# Set the coin_id column as the index
df_scaled_data.set_index("coin_id", inplace=True)

# Display the scaled DataFrame
df_scaled_data.head()


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Scaled DataFrame.

In [329]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 12))

In [330]:
# Create an empty list to store the inertia values
inertia = []


# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for k in k_values:
    kmod = KMeans(n_clusters=k, random_state=0)
    kmod.fit(df_scaled_data)
    inertia.append(kmod.inertia_)
print(inertia)

[287.0, 198.5718180979813, 145.63574772192862, 79.02243535120975, 66.21778006000338, 59.13907002537009, 48.40631449041361, 38.99358798927969, 34.40059545400234, 28.57360256096785, 23.890044355377952]


In [331]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}


# Create a DataFrame with the data to plot the Elbow curve
df_elbow_curve = pd.DataFrame(elbow_data)
df_elbow_curve.head(10)

k     inertia
0  11  287.000000
1  11  198.571818
2  11  145.635748
3  11   79.022435
4  11   66.217780
5  11   59.139070
6  11   48.406314
7  11   38.993588
8  11   34.400595
9  11   28.573603

In [332]:
elbow_curve = df_elbow_curve.hvplot.line(
    x="k",
    y="inertia",
    title="Elbow Curve",
    xlabel="Number of Clusters",
    ylabel="Inertia",
    line_width=2,
    width=600,
    height=300
)

elbow_curve

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Scaled DataFrame

In [333]:
# Initialize the K-Means model using the best value for k
k_means_model = KMeans(n_clusters=4, random_state=1)

In [334]:
# Fit the K-Means model using the scaled DataFrame
k_means_model.fit(market_data_scale)

KMeans(n_clusters=4, random_state=1)

In [335]:
# Predict the clusters to group the cryptocurrencies using the scaled DataFrame
cluster_predictions = k_means_model.predict(df_scaled_data)
# Print the resulting array of cluster values.
print(cluster_predictions)

[2 2 0 0 2 2 2 2 2 0 0 0 0 2 0 2 0 0 2 0 0 2 0 0 0 0 0 0 2 0 0 0 3 2 0 0 1
 0 0 0 0]


/opt/anaconda3/envs/dev/lib/python3.10/site-packages/sklearn/base.py:486: UserWarning: X has feature names, but KMeans was fitted without feature names
  warnings.warn(


In [336]:
# Create a copy of the scaled DataFrame
df_scaled_data_copy = df_scaled_data.copy()

In [337]:
# Add a new column to the copy of the scaled DataFrame with the predicted clusters
df_scaled_data_copy["cluster"] = cluster_predictions

# Display the copy of the scaled DataFrame
df_scaled_data_copy.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  cluster  
coin_id                                            
bitcoin                        -0.251637        2  
ethereum                       -0.199352        2  
tether                         -0.282061        0  
ripple                         -0.295546        0  
bitcoin-cash                   -0.270317        2

In [338]:
# Create a scatter plot using hvPlot by setting
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
plot = df_scaled_data_copy.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d", 
    hover_cols=["coin_id"], 
    title="Cryptocurrency Clusters Price Changes",
    xlabel="24h Price Change",
    ylabel="7d Price Change",
    width=800,
    height=400
)

# Display the scatter plot
plot

:Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [339]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [340]:
# Use the PCA model with `fit_transform` to reduce the original scaled DataFrame
# down to three principal components.
pca_market_data = pca.fit_transform(market_data_scale)

# View the scaled PCA data
pca_market_data

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527],
       [-0.51653377,  1.38837748,  0.80407131],
       [-0.45071134,  0.51769912,  2.84614316],
       [-0.34559977,  0.72943939,  1.47801284],
       [-0.64946792,  0.43216514,  0.60030286],
       [-0.75901394, -0.20119979, -0.21765292],
       [-0.24819846, -1.37625159, -1.46202571],
       [-0.43840762, -0.17533654, -0.6633884 ],
       [-0.69342533, -0.47381462, -0.52759693],
       [ 0.06049915,  2.90940385,  1.49857131],
       [-0.39335243, -0.10819197, -0.01275608],
       [-0.79617564, -0.49440875,  1.08281169],
       [ 0.06407452, -1.26982514, -1.09882928],
       [-0.48901506, -0.73271912, -0.06254323],
       [-0.3062723 ,  0.70341515,  1.71422359],
       [-0.51352775, -0.14280239, -0.65656583],
       [-0.36212044, -0.98691441, -0.728

In [341]:
# Retrieve the explained variance to determine how much information
explained_variance = pca.explained_variance_ratio_
print(explained_variance)

# can be attributed to each principal component.
total_explained_variance = explained_variance.sum()
print(total_explained_variance)

[0.3719856  0.34700813 0.17603793]
0.8950316570309842


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.895

In [342]:
# Create a new DataFrame with the PCA data.
pca_df = pd.DataFrame(
    pca_market_data,
    columns=["PCA1", "PCA2", "PCA3"],
    index=df_scaled_data_copy.index
)
# Copy the crypto names from the original scaled DataFrame
pca_df["coin_id"] = df_scaled_data_copy.index


# Set the coin_id column as index
pca_df.set_index("coin_id", inplace=True)

# Display the scaled PCA DataFrame
pca_df.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the Scaled PCA DataFrame

In [343]:
# Create a list with the number of k-values from 1 to 11
k_values = list(range(1, 12))
k_values


[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]

In [344]:
# Create an empty list to store the inertia values
inertia_pca = []
# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list
for k in k_values:
    kmod = KMeans(n_clusters=k, random_state=0)
    kmod.fit(pca_df)  
    inertia_pca.append(kmod.inertia_)

    print(inertia_pca)


[256.8740855678926]
[256.8740855678926, 168.81189630685233]
[256.8740855678926, 168.81189630685233, 115.91807663320071]
[256.8740855678926, 168.81189630685233, 115.91807663320071, 49.66549665179738]
[256.8740855678926, 168.81189630685233, 115.91807663320071, 49.66549665179738, 41.35731380312857]
[256.8740855678926, 168.81189630685233, 115.91807663320071, 49.66549665179738, 41.35731380312857, 30.79664178704159]
[256.8740855678926, 168.81189630685233, 115.91807663320071, 49.66549665179738, 41.35731380312857, 30.79664178704159, 25.91545691045613]
[256.8740855678926, 168.81189630685233, 115.91807663320071, 49.66549665179738, 41.35731380312857, 30.79664178704159, 25.91545691045613, 21.890936274633233]
[256.8740855678926, 168.81189630685233, 115.91807663320071, 49.66549665179738, 41.35731380312857, 30.79664178704159, 25.91545691045613, 21.890936274633233, 17.720698854473675]
[256.8740855678926, 168.81189630685233, 115.91807663320071, 49.66549665179738, 41.35731380312857, 30.79664178704159, 2

In [345]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_plot = {"k": k_values, "inertia": inertia_pca}
# Create a DataFrame with the data to plot the Elbow curve
df_elbow_plot = pd.DataFrame(elbow_data_plot)

In [346]:
# Plot a line chart with all the inertia values computed with
# the different values of k to visually identify the optimal value for k.
line_chart = df_elbow_plot.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve of Optima; Value for k", 
    xlabel="Number of Clusters",
    ylabel="Inertia",
    line_width=2,
    width=600,
    height=300
)
line_chart

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No, it does not differ from the best k value found using the original data. 

### Cluster Cryptocurrencies with K-means Using the Scaled PCA DataFrame

In [347]:
# Initialize the K-Means model using the best value for k
kmod_pca = KMeans(n_clusters=4, random_state=0)

In [348]:
# Fit the K-Means model using the PCA data
kmod_pca.fit(pca_df)

KMeans(n_clusters=4, random_state=0)

In [349]:
# Predict the clusters to group the cryptocurrencies using the scaled PCA DataFrame
pca_clusters = kmod_pca.predict(pca_df)

# Print the resulting array of cluster values.
pca_clusters

array([2, 2, 0, 0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 0, 2, 0, 0, 2,
       0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 1, 2, 0, 0, 3, 0, 0, 0, 0],
      dtype=int32)

In [350]:
# Create a copy of the scaled PCA DataFrame
pca_df_copy = pca_df.copy()

# Add a new column to the copy of the PCA DataFrame with the predicted clusters
pca_df_copy["cluster"] = pca_clusters

# Display the copy of the scaled PCA DataFrame
pca_df_copy.head()

PCA1      PCA2      PCA3  cluster
coin_id                                            
bitcoin      -0.600667  0.842760  0.461595        2
ethereum     -0.458261  0.458466  0.952877        2
tether       -0.433070 -0.168126 -0.641752        0
ripple       -0.471835 -0.222660 -0.479053        0
bitcoin-cash -1.157800  2.041209  1.859715        2

In [351]:
# Create a scatter plot using hvPlot by setting
# `x="PC1"` and `y="PC2"`.
# Color the graph points with the labels found using K-Means and
# add the crypto name in the `hover_cols` parameter to identify
# the cryptocurrency represented by each data point.
pca_plot = pca_df_copy.hvplot.scatter(
    x="PCA1",  
    y="PCA2",
    by='cluster', 
    hover_cols=["coin_id"], 
    title="Cryptocurrency Clusters in PCA",
    width=600,
    height=300
)

# Display the scatter plot
pca_plot


:NdOverlay   [cluster]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [352]:
# Composite plot to contrast the Elbow curves
df_elbow_curve.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k, 
    width=600,
    height=300
)

:Curve   [k]   (inertia)

In [353]:
df_elbow_plot.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k,
     width=600,
    height=300
)

:Curve   [k]   (inertia)

In [354]:
# Composite plot to contrast the clusters
df_scaled_data_copy.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by='cluster', 
    hover_cols = 'coin_id', 
)

:NdOverlay   [cluster]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

In [355]:
pca_df_copy.hvplot.scatter(
    x="PCA1",  
    y="PCA2",  
    c="cluster", 
    hover_cols=["coin_id"],  
    title="Cryptocurrency Clusters in PCA",
    xlabel="PCA1",
    ylabel="PCA2",
    width=600,
    height=300
)

:Scatter   [PCA1]   (PCA2,cluster,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Using K-means, is more efficient as it reduces to only the most impactful features. The data shows this becauause the clusters become more tightly grouped instead of spread throughout the model.